### ISSUE TITLE
Kafka message bus interacts (unintentionally?) with kafka sink module

### ISSUE THREAD

**Issue by clemper** [spring-projects/spring-xd#1804]
We have encountered an issue when using the kafka message bus along with the kafka sink module.  

The kafka message bus uses KafkaMessageDrivenChannelAdapter from the Spring Integration Kafka project which sets the "kafka_partitionId" message header when consuming from any intermediate message bus topic [here] (https://github.com/spring-projects/spring-integration-kafka/blob/master/src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java#L200).

The kafka sink module uses KafkaProducerMessageHandler from the Spring Integration Kafka project which sets the target partition of the outbound message [here] (https://github.com/spring-projects/spring-integration-kafka/blob/master/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java#L80).

Because the kafka.xml context file does not specify a "partition-id" or "partition-id-expression" attribute on the int-kafka:outbound-channel-adapter element [here] (https://github.com/spring-projects/spring-xd/blob/master/modules/sink/kafka/config/kafka.xml#L12), the outbound target partition is determined by the "kafka_partitionId" message header which is set by the kafka message bus.

Note that almost never is there any direct relationship between and internal Spring XD kafka message bus topic and an external kafka sink topic, so often the number of partitions differ. Also, there is generally no reason to force a correspondence between the message bus topic partition and the external topic.

Does the Spring XD team have any suggestions for working around this issue?  The easiest workaround that I can think is to set the "kafka_paritionId" header to null before the message is sent to the sink; however, I cannot see any straight-forward way to accomplish this.

Thanks,
Chris

**Comment by garyrussell** [comment]
You are correct, in that headers from the `MessageBus` interactions with Kafka should not "leak" into the user (module) space.

Please open a [JIRA Issue](https://jira.spring.io/browse/XD) for this bug.

In the meantime a "quick" work-around would be to add an `<header-filter/>` in a custom version of the sink, upstream of the channel adapter, to remove that header. See the [Spring Integration docs](http://docs.spring.io/spring-integration/reference/html/messaging-transformation-chapter.html#header-filter) for how to configure the filter.

**Comment by mbogoevici** [comment]
Hi @clemper,

Thanks for reporting this, this is not an intended behaviour and is a bug. I've added https://jira.spring.io/browse/XD-3610 to track it. Please use the solution suggested by @garyrussell, we will provide this (or similar behaviour) OOTB in 1.3 M2.

Thanks for reporting this.

**Comment by clemper** [comment]
Based on Gary's suggestion, I created a separate header-filter custom module that simply wraps the org.springframework.integration.transformer.HeaderFilter transformer.  I then created a separate composed module with the header-filter before the kafka sink:

```
module compose --name kafka-out --definition "header-filter --headerNames='kafka_*' --patternMatch=true | sink:kafka"
```

The header-filter module context looks like:

```xml
  <channel id="input" />

  <transformer input-channel="input" output-channel="output">
    <beans:bean class="org.springframework.integration.transformer.HeaderFilter">
      <beans:constructor-arg value="${headerNames}" />
      <beans:property name="patternMatch" value="${patternMatch}" />
    </beans:bean>
  </transformer>

  <channel id="output" />
```

I wired the transformer directly instead of using \<header-filter/\> only so I could make "patternMatch" an module option.

We have this deployed now in our cluster and it seems to resolve the issue.

Thanks,
Chris

**Comment by garyrussell** [comment]
:+1: Good work.

However, `pattern-match` is available on the namespace (default `true`)...

		<xsd:attribute name="pattern-match" type="xsd:boolean" default="true">
			<xsd:annotation>
				<xsd:documentation>
					Boolean flag that specifies whether values provided in 'header-names' should be treated as
					match patterns or literal values. For example header-names='foo*' would mean remove all
					headers that begin with 'foo' including the header named 'foo*'. However if you want to
					treat '*' as literal value setting this flag to FALSE will not perform pattern match and
					the only header that will be removed is the one that is an exact match.
				</xsd:documentation>
			</xsd:annotation>
		</xsd:attribute>

**Comment by clemper** [comment]
The issue with patternMatch is that it is defined as type "xsd:boolean" which is validated when the XML is parsed before the Spring property placeholder handling.  So the XML parser itself complains that "${patternMatch}" is not a boolean.

**Comment by garyrussell** [comment]
Oops - yes; we need to make it a union with `xsd:string` - we've fixed a bunch of those over the years.

I'll raise a JIRA issue; sorry for the noise.
